File: services/ingestion-openmeteo/app/Application/Weather/DTO/WeatherSnapshotData.php
<?php

namespace App\Application\Weather\DTO;

use App\Domain\Weather\Location;

/**
 * Ovaj DTO predstavlja naš interni, normalizovani model vremenskog snapshot-a.
 *
 * Bitno je da application sloj ne zavisi od raw strukture eksternog API-ja.
 * Provider može vratiti šta god želi, ali mi kroz sistem prosleđujemo samo
 * stabilan i jasno definisan DTO.
 */
final readonly class WeatherSnapshotData
{
    public function __construct(
        public string $locationId,
        public string $city,
        public string $country,
        public float $latitude,
        public float $longitude,
        public float $temperatureC,
        public float $windSpeedKmh,
        public int $weatherCode,
        public string $observedAt,
        public string $source,
    ) {
    }

    /**
     * Pretvara DTO u niz podataka pogodan za event payload ili upis u bazu.
     */
    public function toArray(): array
    {
        return [
            'location_id' => $this->locationId,
            'city' => $this->city,
            'country' => $this->country,
            'latitude' => $this->latitude,
            'longitude' => $this->longitude,
            'temperature_c' => $this->temperatureC,
            'wind_speed_kmh' => $this->windSpeedKmh,
            'weather_code' => $this->weatherCode,
            'observed_at' => $this->observedAt,
            'source' => $this->source,
        ];
    }
}

File: services/ingestion-weatherapi/internal/domain/weather/weather_snapshot_data.go
package weather

/*
WeatherSnapshotData predstavlja naš interni, normalizovani model vremenskog snapshot-a.

Ovo je canonical domain model unutar servisa:
- source može biti WeatherAPI danas
- neki drugi API sutra
- ali ostatak sistema radi sa istim oblikom podataka.
*/
type WeatherSnapshotData struct {
	LocationID   string
	City         string
	Country      string
	Latitude     float64
	Longitude    float64
	TemperatureC float64
	WindSpeedKmh float64
	WeatherCode  int
	ObservedAt   string
	Source       string
}

/*
ToMap vraća snapshot kao mapu pogodnu za event payload ili debug ispis.
*/
func (d WeatherSnapshotData) ToMap() map[string]any {
	return map[string]any{
		"location_id":    d.LocationID,
		"city":           d.City,
		"country":        d.Country,
		"latitude":       d.Latitude,
		"longitude":      d.Longitude,
		"temperature_c":  d.TemperatureC,
		"wind_speed_kmh": d.WindSpeedKmh,
		"weather_code":   d.WeatherCode,
		"observed_at":    d.ObservedAt,
		"source":         d.Source,
	}
}
